### PR TITLE
Support client-only mode for NioTransport

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently, NioTransport does not start socket selectors and the
client when the network server setting is set to false. This commit
makes it so that the client will be started even when the network server
is not enabled.

Additionally, it randomly introduces the NioTransport as an option for
the MockTransportClient throughout tests.

### CODE REVIEW

**Comment by yaauie** [comment]
> it randomly introduces the NioTransport as an option for
> the MockTransportClient throughout tests.
> -- [@tbrooks8](https://github.com/elastic/elasticsearch/pull/25839#issue-244803500)

Randomly selecting an alternate codepath in tests seems like a recipe for hard-to-reproduce bugs and builds that succeed only because they randomly failed to exercise code that created a bug; perhaps this option should be parameterised, and _all_ of the tests run through both?

**Comment by nik9000** [comment]
> Randomly selecting an alternate codepath in tests seems like a recipe for hard-to-reproduce bugs and builds that succeed only because they randomly failed to exercise code that created a bug; perhaps this option should be parameterised, and all of the tests run through both?

It *is* parameterize, just, randomly. Elasticsearch uses [this](https://github.com/randomizedtesting/randomizedtesting/wiki) randomized testing framework which allows us to reproduce random tests by starting with the same seed. That is why the seed is in every failure message.

We do have to be very careful though, routing all random decisions through carrotsearch's random source if we are running in test land, but we have the forbiddenapis static analysis to help make sure we do that right.

File: core/src/test/java/org/elasticsearch/client/transport/TransportClientHeadersTests.java [hunk 1]
```diff
@@ -79,10 +82,12 @@ protected Client buildClient(Settings headersSettings, GenericAction[] testedAct
         transportService = MockTransportService.createNewService(Settings.EMPTY, Version.CURRENT, threadPool, null);
         transportService.start();
         transportService.acceptIncomingRequests();
+        String transport = randomBoolean() ? NioTransportPlugin.NIO_TRANSPORT_NAME : MockTcpTransportPlugin.MOCK_TCP_TRANSPORT_NAME;
```
Comment (s1monw) [inline core/src/test/java/org/elasticsearch/client/transport/TransportClientHeadersTests.java]: should we push this into a util method in EsTestCase?

File: test/framework/src/main/java/org/elasticsearch/transport/MockTransportClient.java [hunk 2]
```diff
@@ -37,19 +40,34 @@ public MockTransportClient(Settings settings, Class<? extends Plugin>... plugins
...
+        String transportType = NetworkModule.TRANSPORT_TYPE_SETTING.get(settings);
+        if (settingExists == false || MockTcpTransportPlugin.MOCK_TCP_TRANSPORT_NAME.equals(transportType)) {
+            if (plugins.contains(MockTcpTransportPlugin.class)) {
+                return plugins;
+            } else {
+                plugins = new ArrayList<>(plugins);
+                plugins.add(MockTcpTransportPlugin.class);
+                return plugins;
+            }
+        } else if (NioTransportPlugin.NIO_TRANSPORT_NAME.equals(transportType)) {
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/MockTransportClient.java]: should this also check if the setting exists?

File: test/framework/src/main/java/org/elasticsearch/transport/nio/NioTransportPlugin.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Override
+    public Map<String, Supplier<Transport>> getTransports(Settings settings, ThreadPool threadPool, BigArrays bigArrays,
+                                                          CircuitBreakerService circuitBreakerService,
+                                                          NamedWriteableRegistry namedWriteableRegistry,
+                                                          NetworkService networkService) {
+        Settings settings1;
+        if (NioTransport.NIO_WORKER_COUNT.exists(settings) == false) {
+            // As this is only used for tests right now, limit the number of worker threads.
+            settings1 = Settings.builder().put(settings).put(NioTransport.NIO_WORKER_COUNT.getKey(), 2).build();
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/NioTransportPlugin.java]: ++ what about 1? I mean that should still work no?

**Review by s1monw — CHANGES_REQUESTED** [review]
left some comments Looks Good in general

File: test/framework/src/main/java/org/elasticsearch/transport/nio/NioTransportPlugin.java (hunk 3 above)
Comment (tbrooks8) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/NioTransportPlugin.java]: Yeah. I guess the reason I went with 2, is that we would generally expect the nio transport to be used with multiple selectors. And I thought maybe we would want the testing to always include some level of concurrency? But I can use 1 if you think that is better. Just let me know.

File: test/framework/src/main/java/org/elasticsearch/transport/MockTransportClient.java (hunk 2 above)
Comment (tbrooks8) [inline test/framework/src/main/java/org/elasticsearch/transport/MockTransportClient.java]: I check in the `if` block. If is does not exist, we enter the `if` block and never go to the `else if` block.

File: test/framework/src/main/java/org/elasticsearch/transport/nio/NioTransportPlugin.java (hunk 3 above)
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/NioTransportPlugin.java]: yeah lets stick with 2 we can still randomize it down the road

File: test/framework/src/main/java/org/elasticsearch/transport/MockTransportClient.java (hunk 2 above)
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/MockTransportClient.java]: nevermind I was confused... all is good

**Review by s1monw — APPROVED** [review]
LGTM

**Pull request merged**